src/fn_call.rs

```diff
@@ -151,8 +156,196 @@ pub trait EvalContextExt<'a, 'mir, 'tcx: 'a + 'mir>: crate::MiriEvalContextExt<'
+                let cur_frame = this.cur_frame();
+
+                this.push_stack_frame(
+                    box_me_up_fn,
+                    box_me_up_mir.span,
+                    box_me_up_mir,
+                    Some(temp_ptr.into()),
+                    StackPopCleanup::None { cleanup: true }
+                )?;
+
+                let mut args = this.frame().mir.args_iter();
+                let arg_0 = this.eval_place(&mir::Place::Base(mir::PlaceBase::Local(args.next().unwrap())))?;
+                this.write_scalar(data_ptr, arg_0)?;
+
+                // Step through execution of 'box_me_up'
+                // We know that we're finished when our stack depth
+                // returns to where it was before.
+                //
+                // Note that everything will get completely screwed up
+                // if 'box_me_up' panics. This is fine, since this
+                // function should never panic, as it's part of the core
+                // panic handling infrastructure
+                //
+                // Normally, we would just let Miri drive
+                // the execution of this stack frame.
+                // However, we need to access its return value
+                // in order to properly unwind.
+                //
+                // When we 'return' from '__rustc_start_panic',
+                // we need to be executing the panic catch handler.
+                // Therefore, we take care all all of the unwinding logic
+                // here, instead of letting the Miri main loop do it
+                while this.cur_frame() != cur_frame {
+                    this.step()?;
+                }
+
+                // 'box_me_up' has finished. 'temp_ptr' now holds
+                // a '*mut (dyn Any + Send)'
+                // We want to split this into its consituient parts -
+                // the data and vtable pointers - and store them back
+                // into the panic handler frame
+                let real_ret = this.read_immediate(temp_ptr.into())?;
+                let real_ret_data = real_ret.to_scalar_ptr()?;
+                let real_ret_vtable = real_ret.to_meta()?.expect("Expected fat pointer");
+
+                // We're in panic unwind mode. We pop off stack
+                // frames until one of two things happens: we reach
+                // a frame with 'catch_panic' set, or we pop of all frames
+                //
+                // If we pop off all frames without encountering 'catch_panic',
+                // we exit.
+                //
+                // If we encounter 'catch_panic', we continue execution at that
+                // frame, filling in data from the panic
+                //
+                unwind_stack(this, real_ret_data, real_ret_vtable)?; 
+
+                this.memory_mut().deallocate(temp_ptr.to_ptr()?, None, MiriMemoryKind::UnwindHelper.into())?;
+                this.dump_place(*dest.expect("dest is None!"));
```

Comment:
This code is funny. When exactly does this get executed? Looks like *after* stack unwinding, like when the panic got caught but before whoever caught it gets to run? That's really weird control flow! And if unwinding was more integrated into the engine [the way I think it should be](https://github.com/rust-lang/rust/pull/60026#discussion_r281015251), this would not even be possible.

Is there an equivalent to this in the real implementation?